### PR TITLE
net: lib: nrf_cloud: add location and device message send to rest lib

### PR DESCRIPTION
**Pull request by jayteemo**
Add functions to send location (GPS) message and general device
messages to nRF Cloud using REST.
CIA-415

Signed-off-by: Justin Morton <justin.morton@nordicsemi.no>

### CODE REVIEW

**Comment by jayteemo** [comment]
draft for no... need to test with backend fix that has been recently pushed to dev.

**Review by plskeggs — APPROVED** [review]
Looks good.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1177]: You compare cJSON_AddStringToObjectCS result against NULL here, but later compare against cJSON_True -- is that intentional?

**Review by glarsennordic — APPROVED** [review]
Looks good AFAICT

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1177]: not intentional, fixed.

File: include/net/nrf_cloud_rest.h
Comment (jayteemo) [reply include/net/nrf_cloud_rest.h:302]: external documentation links can be found in the various readme files... in general, i don't think we do any links in source code.  i'm not sure if there are any official rules about this though.

Comment (jtguggedal) [reply include/net/nrf_cloud_rest.h:302]: I also wondered about this for another API. I dropped the link in .h docs, as the API documentation is anyway presented on the same page as the library docs from the .rst file and links can be put there: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/networking/nrf_cloud_rest.html#id3

File: include/net/nrf_cloud_rest.h
Comment (jtguggedal) [inline include/net/nrf_cloud_rest.h:320]: It could also be nice to have the option to send PVT data, but I guess there needs to be cloud-side support first

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1244]: Do we need calloc or can we get away with malloc? Buffer can be null-terminated at line 1257, after the overflow check. Not a big deal, but would shave off a few CPU cycles.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1264]: Nit: This can be moved to the top of the function with the other variable definitions

File: include/net/nrf_cloud_rest.h
Comment (jayteemo) [reply include/net/nrf_cloud_rest.h:320]: nrf cloud supports the nmea sentences listed here:
https://github.com/nRFCloud/application-protocols/tree/v1/schemas/deviceToCloud/gps

File: include/net/nrf_cloud_rest.h
Comment (jayteemo) [reply include/net/nrf_cloud_rest.h:302]: this is a non-NCS link: https://docs.nrfcloud.com/Reference/DeviceMessages/#sending-device-messages
i am hesitant to add it to the .h file simply for maintenance purposes.
but also, as noted in the linked doc page, _any_ json schema is accepted by the cloud... so i've removed the formatting comment.

i'll add the link if you want it in there.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c]: done... updated the other occurrences as well.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1264]: `auth_hdr` is initially NULL and allocated just a couple lines above in `generate_auth_header`.

File: include/net/nrf_cloud_rest.h
Comment (glarsennordic) [reply include/net/nrf_cloud_rest.h]: Honestly, go with what you think is best. My personal instinct would be to add the link, but maintenance is a completely valid concern. I think you are better equipped than me to make the cost benefit analysis on that

**Review by rlubos — APPROVED** [review]

**Comment by jayteemo** [comment]
made a few changes after testing and removed draft state.

**Review by plskeggs — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1322]: Since you call cJSON_AddRawToObjectCS with json_msg, is it possible that this delete call will recursively delete the passed in json_msg before the caller is expecting it? Could this lead to a double deletion? Or are these objects reference counted by cJSON?

I'm not 100% certain on how cJSON memory management is structured.

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1322]: I see some mention of reference counting in some cJSON sources, but I don't know whether AddRawToObjectCS will increment the reference count, nor if cJSON_Delete even respects the reference count.

However, if you find that deletion/addition do not increment/decrement some reference counter, this presents itself as a potential solution:
https://cjson.docsforge.com/master/api/cJSON_CreateObjectReference/

It's also totally possible that I'm just misreading your code, and that this isn't a problem to begin with

**Review by glarsennordic — APPROVED** [review]
Looks good to the extent that I am able to judge, assuming the double delete thing isn't an issue!

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c:1322]: Oh, yeah, I totally misunderstand. Passed in json_msg is a raw char*, and the cJSON_AddRawToObjectCS function just creates a reference.

This is a non-issue

File: include/net/nrf_cloud_rest.h
Comment (melwee01) [reply include/net/nrf_cloud_rest.h]: Maintenance is a big concern as we've had ongoing organization work that might affect links. That's not a blocker, but we need to keep in communication to make sure it's not unexpectedly broken.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (melwee01) [inline doc/nrf/releases/release-notes-changelog.rst:333]:
```suggestion
  * Added :c:func:`nrf_cloud_rest_send_device_message` function that sends a JSON message to nRF Cloud using the `SendDeviceMessage` endpoint.
```

Consulted the docs team involved with release note style. API operation/endpoint names like this should be in backticks to match how they appear in the nRF Cloud docs.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (melwee01) [inline doc/nrf/releases/release-notes-changelog.rst:334]:
```suggestion
  * Added :c:func:`nrf_cloud_rest_send_location` function that sends an NMEA sentence to nRF Cloud as a GPS-type device message.
```

**Review by melwee01 — DISMISSED** [review]

File: doc/nrf/releases/release-notes-changelog.rst
Comment (melwee01) [reply doc/nrf/releases/release-notes-changelog.rst:334]: Cleared my 'requested changes'. I approve once these small changes are put in, and don't want to block merging since it'll likely be merged when I'm not here due to time zones.

**Comment by rlubos** [comment]
@jayteemo Please apply the suggestions and rebase the PR (conflicts...).

File: doc/nrf/releases/release-notes-changelog.rst
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:348]:
```suggestion
  * Added :c:func:`nrf_cloud_rest_send_device_message` function that sends a JSON message to nRF Cloud using the ``SendDeviceMessage`` endpoint.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Review by peknis — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst
Comment (umapraseeda) [reply doc/nrf/releases/release-notes-changelog.rst:348]: Let's fix this when the changelog is converted into the release notes. I shall do that.

**Pull request merged**
